### DISCUSSION TITLE
Why are some dataclass configurations not supported?

### DISCUSSION THREAD

**Discussion by lonelyteapot** [sqlalchemy/sqlalchemy#8345] [Usage Questions] [answered]
I'm looking into SQLAlchemy 2.0.0b1 and the new dataclasses mapping feature.

Here's a quote from [the docs](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#class-level-feature-configuration):

> Support for dataclasses features is partial. Currently supported are the `init`, `repr`, `eq`, `order` and `unsafe_hash` features. Currently not supported are the `frozen`, `slots`, `match_args`, and `kw_only` features.

Is there any particular reason they're not supported? If not, are they being worked on currently? I couldn't find any issues nor PRs regarding this. I'm not that familiar with the project, so I hesitate filing an issue myself.

My use case needs `kw_only` enabled. I'd like to declare models like this:

```python
class Foo(MappedAsDataclass, Base, kw_only=True):
    id: Mapped[UUID] = mapped_column(UUID, primary_key=True, default_factory=uuid4)
    name: Mapped[str] = mapped_column()
```

Notice that a default-initialized field `id` comes before a not-default-initialized field `name`. This is not possible without `kw_only=True` because of the conflict of positional arguments in the generated `__init__` method.

**Comment by CaselIT** [comment, marked as answer]
> Is there any particular reason they're not supported?

I think the main things that are unlikely to ever get supported are slot and frozen, the first because sqlalchemy requires dict, the other is because it's likely to interfere with the usual orm think.

Regarding the rest they were deemed not prioritaty but should not cause much issues, since sqlalchemy does not use the generated __init__ and match are ignored by sqlalchemy.
They are not been worked on, but they should be quite straight forward to implement and you have a good use case for kw_only

**Reply by CaselIT** [reply]
created https://github.com/sqlalchemy/sqlalchemy/issues/8346

**Reply by CaselIT** [reply]
another thing is that they were added in later version of dataclasses.

but I think we can just let dataclass error of one tries to use kw_only on 3.9

**Comment by zzzeek** [comment]
we can in theory support all of those options, even slots (although it would be a lot of work with questionable benefits)  Slight exception to "match_args" as Python's documentation doesnt seem to indicate in any way what this parameter means.  I wanted to get the dataclasses feature out the door with a reasonable set of features and as people want the other features, they can be added.

**Reply by CaselIT** [reply]
Match args is used for the new match construct.
I've implement kw_only and match_args. Seems very straight forward.

**Reply by CaselIT** [reply]
See for info on match args https://docs.python.org/3/reference/datamodel.html?highlight=__match_args__#object.__match_args__
